pathfinding_sdf/pathfinding_sdf.c

```diff
@@ -0,0 +1,758 @@
+*        this could again be done using the SDF values to determine if a node can be removed
+*        without clipping through walls
+*
+********************************************************************************************/
+
+#include "raylib.h"
+#include "raymath.h"
+
+#include <stddef.h> // Required for: NULL
+#include <math.h> // Required for: abs
+
+typedef struct PathfindingNode
+{
+    int x, y;
+    int fromX;
+    int fromY;
+    int score;
+} PathfindingNode;
+
+typedef struct NeighborOffset
+{
+    int x, y;
+    int distance;
+} NeighborOffset;
+
+typedef struct Agent
+{
+    int startX, startY;
+    int targetX, targetY;
+    int wallFactor;
+    int unitSize;
+    PathfindingNode *path;
+    int pathCount;
+    PathfindingNode *map;
+    Vector2* icon;
+    Color color;
+    int iconCount;
+    float walkedPathDistance;
+} Agent;
+
+typedef struct AppState
+{
+    int visualizeMode;
+    int randomizeBlocks;
+    char paintMode;
+    int updateSDF;
+    int sdfFunction;
+    int jumpingEnabled;
+    int cellX, cellY;
+    Agent rat;
+    Agent cat;
+} AppState;
+
+// a simple cat face that can be drawn as a triangle fan
+Vector2 catFace[] = {
+    {0.0f, 1.0f},
+    {0.7f, 1.0f},
+    {1.0f, 0.7f},
+    {1.0f, -1.0f},
+    {0.5f, -0.6f},
+    {-0.5f, -0.6f},
+    {-1.0f, -1.0f},
+    {-1.0f, 0.7f},
+    {-0.7f, 1.0f},
+};
+
+// a simple rat face that can be drawn as a triangle fan
+Vector2 ratFace[] = {
+    {0.0f, 1.0f},
+    {0.3f, 0.9f},
+    {0.8f, -0.2f},
+    {1.0f, -0.8f},
+    {0.8f, -1.0f},
+    {0.5f, -1.0f},
+    {0.3f, -0.7f},
+    {0.0f, -0.8f},
+    {-0.3f, -0.7f},
+    {-0.5f, -1.0f},
+    {-0.8f, -1.0f},
+    {-1.0f, -0.8f},
+    {-0.8f, -0.2f},
+    {-0.3f, 0.9f},
+};
+
+
+const Color gridColor = { 200, 200, 200, 40 };
+const Color cellHighlightColor = { 200, 0, 0, 80 };
+const float movementSpeed = 3.0f;
+
+const int gridWidth = 80;
+const int gridHeight = 45;
+const int cellSize = 10;
+
+char* blockedCells = NULL;
+char* sdfCells = NULL;
+
+// lookup table for cheap square root calculation
+int isqrt[256] = {0};
+
+// various offsets and distances for jumping nodes during pathfinding
+NeighborOffset neighborOffsets[20*20] = {0};
+int neighborOffsetCount = 0;
+
+Agent Agent_init(int x, int y, int size, int targetX, int targetY, int wallFactor, Vector2* icon, int iconCount, Color color)
+{
+    Agent agent;
+    agent.unitSize = size;
+    agent.startX = x;
+    agent.startY = y;
+    agent.targetX = targetX;
+    agent.targetY = targetY;
+    agent.pathCount = 0;
+    agent.wallFactor = wallFactor;
+    agent.path = (PathfindingNode *)MemAlloc(gridWidth * gridHeight * sizeof(PathfindingNode) * 4);
+    agent.map = (PathfindingNode *)MemAlloc(gridWidth * gridHeight * sizeof(PathfindingNode));
+    agent.icon = icon;
+    agent.color = color;
+    agent.iconCount = iconCount;
+    agent.walkedPathDistance = 0.0f;
+    return agent;
+
+}
+
+void Agent_findPath(Agent *agent, int enableJumping)

```

Comment:
this function is huge, can it be broken up more? maybe in places where you have comments, that should be a sub function for readability?

I fear that most people will not be able to follow this code.